pkg/engine/lifecycletest/resource_hooks_test.go

```diff
@@ -0,0 +1,394 @@
+			},
+		})
+		assert.NoError(t, err)
+		return nil
+	})
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF},
+	}
+	p.Steps = []lt.TestStep{{Op: Update}}
+	snap := p.Run(t, nil)
+
+	assert.True(t, hookCalled)
+	assert.Len(t, snap.Resources, 2)
+	assert.Equal(t, snap.Resources[0].URN.Name(), "default")
+	assert.Equal(t, snap.Resources[1].URN.Name(), "resA")
+}
+
+func TestResourceHookDryRun(t *testing.T) {
+	t.Parallel()
+
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{}, nil
+		}),
+	}
+
+	hookTrueCalledOnDryRun := false
+	hookTrueCalled := false
+	hookFalseCalled := false
+
+	programF := deploytest.NewLanguageRuntimeF(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		callbacks, err := deploytest.NewCallbacksServer()
+		require.NoError(t, err)
+		defer func() { require.NoError(t, callbacks.Close()) }()
+
+		// Create hook that runs on a dry run (and non-dry run)
+		funTrue := func(ctx context.Context, urn resource.URN, id resource.ID, inputs resource.PropertyMap,
+			outputs resource.PropertyMap,
+		) error {
+			assert.Equal(t, urn, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resA"))
+			if info.DryRun {
+				hookTrueCalledOnDryRun = true
+			} else {
+				hookTrueCalled = true
+			}
+			return nil
+		}
+		myHookTrue, err := deploytest.NewHook(monitor, callbacks, "myHookTrue", funTrue, true)
+		require.NoError(t, err)
+
+		// Create hook that does not run on a dry run
+		funFalse := func(ctx context.Context, urn resource.URN, id resource.ID, inputs resource.PropertyMap,
+			outputs resource.PropertyMap,
+		) error {
+			assert.Equal(t, urn, resource.URN("urn:pulumi:test::test::pkgA:m:typA::resA"))
+			if info.DryRun {
+				assert.Fail(t, "The hook should not have been called")
+			} else {
+				hookFalseCalled = true
+			}
+			return nil
+		}
+		myHookFalse, err := deploytest.NewHook(monitor, callbacks, "myHookFalse", funFalse, false)
+		require.NoError(t, err)
+
+		// Register a resource with both hooks
+		_, err = monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			ResourceHookBindings: deploytest.ResourceHookBindings{
+				AfterCreate: []*deploytest.ResourceHook{myHookTrue, myHookFalse},
+			},
+		})
+		assert.NoError(t, err)
+		return nil
+	})
+	hostF := deploytest.NewPluginHostF(nil, nil, programF, loaders...)
+
+	p := &lt.TestPlan{
+		Options: lt.TestUpdateOptions{T: t, HostF: hostF},
+	}
+	p.Steps = []lt.TestStep{{Op: Update}}
+	snap := p.Run(t, nil)
+
+	assert.True(t, hookTrueCalledOnDryRun, "hook true should have been called on dry run")
+	assert.True(t, hookTrueCalled, "hook true should have been called")
+	assert.True(t, hookFalseCalled, "hook false should have been called")
+	assert.Len(t, snap.Resources, 2)
```

Comment:
```suggestion
	require.Len(t, snap.Resources, 2)
```

If this is not true, subsequent steps might panic.